packages/tools/src/tools/segmentation/strategies/fillSphere.ts

```diff
@@ -1,136 +1,142 @@
+    const [topLeftCanvas, bottomRightCanvas] =
+      getCanvasEllipseCorners(canvasCoordinates);
+
+    // 2. Find the extent of the ellipse (circle) in IJK index space of the image
+    const topLeftWorld = viewport.canvasToWorld(topLeftCanvas);
+    const bottomRightWorld = viewport.canvasToWorld(bottomRightCanvas);
+    // This will be 2d, now expand to 3d
+    const diameters = topLeftWorld.map((left, index) =>
+      Math.abs(bottomRightWorld[index] - left)
+    );
+    const radius = Math.max(...diameters) / 2;
+    // Make 3d sphere
+    topLeftWorld.forEach((left, index) => {
+      const right = bottomRightWorld[index];
+      if (left === right) {
+        topLeftWorld[index] = left - radius;
+        bottomRightWorld[index] = left + radius;
       }
+    });
+
+    const ellipsoidCornersIJK = [
+      <Types.Point3>transformWorldToIndex(segmentationImageData, topLeftWorld),
+      <Types.Point3>(
+        transformWorldToIndex(segmentationImageData, bottomRightWorld)
+      ),
+    ];
+
+    segmentationVoxelValue.boundsIJK = getBoundingBoxAroundShape(
+      ellipsoidCornersIJK,
+      segmentationVoxelValue.dimensions
+    );
 
-      if (
-        isWithinThreshold(index, imageScalarData, strategySpecificConfiguration)
-      ) {
-        segmentationScalarData[index] = segmentIndex;
-        scalarIndex.push(index);
-      }
-    };
-  } else {
-    callback = ({ index, value }) => {
-      if (segmentsLocked.includes(value)) {
-        return;
-      }
-      segmentationScalarData[index] = segmentIndex;
-      scalarIndex.push(index);
+    // using circle as a form of ellipse
+    const ellipseObj = {
+      center: center as Types.Point3,
+      xRadius: Math.abs(topLeftWorld[0] - bottomRightWorld[0]) / 2,
+      yRadius: Math.abs(topLeftWorld[1] - bottomRightWorld[1]) / 2,
+      zRadius: Math.abs(topLeftWorld[2] - bottomRightWorld[2]) / 2,
     };
-  }
-
-  pointInSurroundingSphereCallback(
-    segmentationImageData,
-    [points[0], points[1]],
-    callback,
-    viewport as Types.IVolumeViewport
-  );
-
-  const dimensions = segmentationImageData.getDimensions();
-
-  let sliceArray;
-  if (isVolumeSegmentation(operationData)) {
-    // Since the scalar indexes start from the top left corner of the cube, the first
-    // slice that needs to be rendered can be calculated from the first mask coordinate
-    // divided by the zMultiple, as well as the last slice for the last coordinate
-    const zMultiple = dimensions[0] * dimensions[1];
-    const minSlice = Math.floor(scalarIndex[0] / zMultiple);
-    const maxSlice = Math.floor(
-      scalarIndex[scalarIndex.length - 1] / zMultiple
-    );
-    sliceArray = Array.from(
-      { length: maxSlice - minSlice + 1 },
-      (v, k) => k + minSlice
-    );
-    triggerSegmentationDataModified(operationData.volumeId, sliceArray);
-  }
-}
+
+    imageVoxelValue.isInObject = (pointLPS /*, pointIJK */) =>
+      pointInEllipse(ellipseObj, pointLPS);
```

Comment:
I'm pretty sure you are reducing the performance here. Previously, we had the `pointInSurroundingSphereCallback`, which was quite efficient in taking into account the camera, etc. Why don't I see that anymore in your changes?

I tried the local demo versus the deployed demo and noticed the performance issue in the sphere brush as well.

Reply:
That is a generic optimization I plan on making for both sphere and circle annotations as part of the 3.x implementation.  What I plan on is a generic set of calls to determine bounds along a given dimension, so that if you provide bounds for zero, one or two of i,j,k, you can get the bounds for the remaining element.  Then, allow one to set it for everything outside, and everything inside.
Probably will end up with 3 versions of the iteration in VoxelValue, each optimized for different situations.